pattern/constraint/value/AssignmentConstraint.java

```diff
@@ -0,0 +1,622 @@
+            } else if (expr.isOperation()) {
+                Pair<com.vaticle.typeql.lang.pattern.constraint.ValueConstraint.Assignment.Expression, com.vaticle.typeql.lang.pattern.constraint.ValueConstraint.Assignment.Expression> operands = expr.asOperation().operands();
+                return new Operation(expr.asOperation().operator(), build(operands.first(), registry), build(operands.second(), registry));
+            } else if (expr.isFunction()) {
+                List<Expression> args = new ArrayList<>();
+                iterate(expr.asFunction().arguments()).map(argExpr -> build(argExpr, registry)).forEachRemaining(args::add);
+                return new Function(expr.asFunction().symbol(), args);
+            } else throw TypeDBException.of(ILLEGAL_STATE);
+        }
+
+        public void compiledExpression(com.vaticle.typedb.core.traversal.expression.Expression<?> compiled) {
+            this.compiled = compiled;
+        }
+
+        public com.vaticle.typedb.core.traversal.expression.Expression<?> compiledExpression() {
+            return compiled;
+        }
+
+
+        public boolean isOperation() {
+            return false;
+        }
+
+        public boolean isFunction() {
+            return false;
+        }
+
+        public boolean isThingVar() {
+            return false;
+        }
+
+        public boolean isValueVar() {
+            return false;
+        }
+
+        public boolean isConstant() {
+            return false;
+        }
+
+        public Operation asOperation() {
+            throw TypeDBException.of(INVALID_CASTING, className(this.getClass()), className(Operation.class));
+        }
+
+        public Function asFunction() {
+            throw TypeDBException.of(INVALID_CASTING, className(this.getClass()), className(Function.class));
+        }
+
+        public ThingVar asThingVar() {
+            throw TypeDBException.of(INVALID_CASTING, className(this.getClass()), className(ThingVar.class));
+        }
+
+        public ValueVar asValueVar() {
+            throw TypeDBException.of(INVALID_CASTING, className(this.getClass()), className(ValueVar.class));
+        }
+
+        public Constant<?> asConstant() {
+            throw TypeDBException.of(INVALID_CASTING, className(this.getClass()), className(Constant.class));
+        }
+
+        @Override
+        public abstract String toString();
+
+        @Override
+        public abstract boolean equals(Object o);
+
+        @Override
+        public abstract int hashCode();
+
+        public abstract Expression clone(Map<Identifier, Variable> argMap);
+
+        public static class Operation extends Expression {
+
+            private final TypeQLToken.Expression.Operation op;
+            private final Expression a;
+            private final Expression b;
+            private final int hash;
+
+            public Operation(TypeQLToken.Expression.Operation op, Expression a, Expression b) {
```

Comment:
pick consistent naming scheme throughout the expression-realted changes. Operations should always use first/second, or lhs/rhs, etc.